File: Graphs/Vertex.java
package Graphs;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * The Vertex class represents a node in a graph. It holds details about
 * the connections to other vertices and can store additional metadata.
 */
public class Vertex {
    // Unique identifier for the vertex. Using String to allow flexible vertex naming.
    private String id;

    // List to store adjacent vertices. This simplifies the management of edges in undirected graphs.
    private List<Vertex> adjacencyList;

    // A map to hold arbitrary metadata for the vertex. This allows extending vertex properties dynamically.
    private Map<String, Object> properties;

    /**
     * Constructor for creating a new Vertex with a given identifier.
     * @param id The unique identifier for the vertex.
     */
    public Vertex(String id) {
        this.id = id;
        this.adjacencyList = new ArrayList<>();
        this.properties = new HashMap<>();
    }

    /**
     * Adds an adjacent vertex to the connectivity list. This method ensures that
     * the graph structure can be built by adding edges from this vertex to others.
     * @param vertex The vertex to be added as adjacent.
     */
    public void addAdjacentVertex(Vertex vertex) {
        if (!adjacencyList.contains(vertex)) { // Prevents duplicate edges
            adjacencyList.add(vertex);
        }
    }

    /**
     * Adds or updates a property to this vertex's metadata.
     * @param key The key under which the property is stored.
     * @param value The value of the property.
     */
    public void addProperty(String key, Object value) {
        properties.put(key, value);
    }

    /**
     * Retrieves a property value by key.
     * @param key The key of the property to retrieve.
     * @return The value of the property, or null if not found.
     */
    public Object getProperty(String key) {
        return properties.get(key);
    }

    /**
     * Returns the unique identifier of this vertex.
     * @return The identifier of the vertex.
     */
    public String getId() {
        return id;
    }

    /**
     * Returns a list of all vertices directly connected to this vertex.
     * @return A list of adjacent vertices.
     */
    public List<Vertex> getAdjacencyList() {
        return adjacencyList;
    }
}

File: Graphs/Vertex.py
class Vertex:
    def __init__(self, id):
        """
        Initialize a new Vertex with a unique identifier.

        :param id: The unique identifier for the vertex.
        """
        self.id = id
        self.adjacency_list = []  # List to store adjacent vertices
        self.properties = {}      # Dictionary to store properties

    def add_adjacent_vertex(self, vertex):
        """
        Add a vertex to the adjacency list of this vertex.

        :param vertex: The Vertex instance to be added as adjacent.
        """
        if vertex not in self.adjacency_list:
            self.adjacency_list.append(vertex)

    def add_property(self, key, value):
        """
        Adds or updates a property associated with this vertex.

        :param key: The key for the property.
        :param value: The value of the property.
        """
        self.properties[key] = value

    def get_property(self, key):
        """
        Retrieves the property value associated with the specified key.

        :param key: The key of the property to retrieve.
        :return: The value of the property, or None if the key does not exist.
        """
        return self.properties.get(key)

    def __str__(self):
        """
        Return a string representation of the vertex and its connections for easier debugging.
        Lists only the IDs of adjacent vertices.
        """
        adjacent_ids = [v.id for v in self.adjacency_list]  # Get IDs of adjacent vertices
        return f"Vertex({self.id}) --> {adjacent_ids}"

# Example of using the Vertex class
if __name__ == "__main__":
    # Create vertices
    v1 = Vertex("1")
    v2 = Vertex("2")
    v3 = Vertex("3")

    # Add edges
    v1.add_adjacent_vertex(v2)
    v1.add_adjacent_vertex(v3)

    # Print vertex
    print(v1)
